src/OpenTelemetry.Instrumentation.Process/ProcessMetrics.cs

```diff
@@ -18,14 +18,15 @@ internal sealed class ProcessMetrics
 
     private static readonly Meter MeterInstance = new(MeterName, Assembly.GetPackageVersion());
 
+    private static readonly Diagnostics.Process Process = Diagnostics.Process.GetCurrentProcess();
+
     static ProcessMetrics()
     {
         MeterInstance.CreateObservableUpDownCounter(
             "process.memory.usage",
             () =>
             {
-                using var process = Diagnostics.Process.GetCurrentProcess();
-                return process.WorkingSet64;
+                return Process.WorkingSet64;
```

Comment:
You are right @haipz probably the class will need some redesign, e.g. maybe the ProcessInstrumentationOptions could accept a refresh rate at which we could refresh the Process. I will try to get ahold of @Yun-Ting  to discuss this.

Reply:
An option to this is using Environment.WorkingSet which has no perf implications and displays the same data with minimal change
<img width="1028" alt="image" src="https://github.com/user-attachments/assets/8549369b-c6df-4b9b-9f08-f43f07cd6e4f">

Reply:
Hi @danespinosa, I've gone down the cache path before. Check out the discussions here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/718.

Reply:
Thanks for sharing @Yun-Ting i see that your implementation got updated intending to solve an issue but it really didn't solve the issue perse. Here is the commit [[Instrumentation.Process, Resources.Process] Properly disposes of Sys… · open-telemetry/opentelemetry-dotnet-contrib@55c0dfb](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/commit/55c0dfbc759f0a13fa00e298c5452726b9aaac32#diff-c76f8af700ddf621d4b8f4ba6f3483bf06a56e3ea60a9574d80a9e2cac3584f3R9) and here is the issue that explains at the bottom that actually the issue persists [Properly dispose instances of System.Diagnostics.Process class · Issue #2100 · open-telemetry/opentelemetry-dotnet-contrib](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/2100).

My 2 suggestions and 2 cents, could we revert that PR and bring back your implementation or can we at least update the process.WorkingSet64 call to instead calling Environment.WorkingSet? This will bring huge gains
<img width="1715" alt="image" src="https://github.com/user-attachments/assets/073e50e2-3142-4929-8da9-47306d4c3b49">

Reply:
@noahfalk  I was chatting with @Yun-Ting regarding the perf or this library and we were wondering if we could get your thoughts about this change.

Reply:
This seems like a fine change to me.

Looking at #718 I see that it relied on some benchmarking in #717. The benchmarking tested various ways of running the Refresh() or GetCurrentProcess() in isolation and then #718 extrapolated that if the calls were fast in isolation then it would also be fast in the process metrics scenarios where we invoke various property getters on the Process object afterwards. Unforetunately that isn't the case. The property getters populate data lazily if it isn't already populated and that winds up being the expensive part of the operation. Calling GetCurrentProcess() returns a process object where the values haven't been initialized and Refresh() only invalidates the cached data but doesn't fetch new data. Calling GetCurrentProcess() or Refresh() repeatedly costs very little. Also calling WorkingSet on an initialized Process object costs very little. But the first call to WorkingSet after Refresh() or GetCurrentProcess() is comparatively very expensive:

```
| RefreshOnly                    |         2.017 ns |      0.0305 ns |       0.0270 ns |
| GetCurrentProcessOnly          |        47.814 ns |      0.9146 ns |       1.1567 ns |
| WorkingSetOnly                 |         1.110 ns |      0.0231 ns |       0.0216 ns |
| RefreshAndWorkingSet           | 3,241,750.985 ns | 58,170.7094 ns | 109,258.7011 ns |
| GetCurrentProcessAndWorkingSet | 3,243,749.014 ns | 63,792.7044 ns |  75,940.6834 ns |
```

Benchmark code:
```C#
    public class ProcessBenchmarks
    {
        private static Process s_process = Process.GetCurrentProcess();


        [Benchmark]
        public void RefreshOnly()
        {
            s_process.Refresh();
        }

        [Benchmark]
        public Process GetCurrentProcessOnly()
        {
            return Process.GetCurrentProcess();
        }

        [Benchmark]
        public long WorkingSetOnly()
        {
            return s_process.WorkingSet64;
        }

        [Benchmark]
        public long RefreshAndWorkingSet()
        {
            s_process.Refresh();
            return s_process.WorkingSet64;
        }

        [Benchmark]
        public long GetCurrentProcessAndWorkingSet()
        {
            return Process.GetCurrentProcess().WorkingSet64;
        }
    }
```

In the .NET source [here](https://source.dot.net/#System.Diagnostics.Process/System/Diagnostics/Process.cs,978) this is the cache check. If processInfo == null the property getter will take milliseconds, if its already cached the getter will be nanoseconds. Hope that helps!

Reply:
This change all good to merge then?

Reply:
My apologies, I have been busy with other things and this took second priority, I'll be working on not only updating the method to get the WorkingSet but also the Memory VirtualMemorySize64. Then in other PRs i'll be improving CPU and ThreadCount and by bringing the logic from the [.NET metrics](https://github.com/dotnet/runtime/blob/8975db4b2b8b671de5a6334da6d2c7bfecdfc7e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RuntimeMetrics.cs#L4) hopefully it's not too hard :)